File: async/omniparallel/py/async.py
#!/usr/bin/env python3

import aiohttp
import asyncio
import sys
import threading
import time

from typing import List


COUNT: int = 500


async def fetchPage(url: str) -> int:
    async with aiohttp.request('GET', url) as resp:
        body = await resp.text()
    return len(str(body))


async def runSync(url: str):
    print("Starting sync...")
    start_ns = time.time_ns()
    responses: List = [await fetchPage(url) for i in range(COUNT)]
    delta_ms = (time.time_ns() - start_ns) / 1000000
    bytes = sum(responses)
    print(f"Sync: {bytes} bytes in {delta_ms}ms")


async def runAsync(url: str):
    print("Starting async...")
    start_ns = time.time_ns()
    tasks = [fetchPage(url) for i in range(COUNT)]
    responses = await asyncio.gather(*tasks)
    delta_ms = (time.time_ns() - start_ns) / 1000000
    bytes = sum(responses)
    print(f"Async: {bytes} bytes in {delta_ms}ms")


async def main():
    if len(sys.argv) < 2:
        myname = sys.argv[0]
        print(f"Usage: {myname} URL")
        sys.exit(1)

    url = sys.argv[1]

    threads = (
        threading.Thread(target=lambda: asyncio.run(runSync(url))),
        threading.Thread(target=lambda: asyncio.run(runAsync(url))),
    )

    print("Starting threads...")
    [t.start() for t in threads]

    print("Joining threads...")
    [t.join() for t in threads]


if __name__ == '__main__':
    asyncio.run(main())

File: async/omniparallel/ts/async.ts
import path from "node:path";
import process from "node:process";
import { firstValueFrom, from, merge, Observable } from "rxjs";
import { reduce } from "rxjs/operators";

type SomeOperation = (url: string, count: number) => Promise<number>;

async function fetchEndpoint(url: string): Promise<number> {
  // start the fetch (header and first chunk only)
  const resp = await fetch(url);
  // ensure full text (may be multiple chunks)
  const text = await resp.text();
  return text.length;
}

async function runSync(url: string, count: number) {
  // invokes async function synchronously {count} times in sequence
  let bytes: number = 0;
  for (let i = 0; i < count; ++i) {
    bytes += await fetchEndpoint(url);
  }
  return bytes;
}

async function runAsync(url: string, count: number): Promise<number> {
  // starts {count} async operations concurrently, collecting their promises
  const promises: Array<Promise<number>> = [];
  for (let i = 0; i < count; ++i) {
    promises.push(fetchEndpoint(url));
  }
  // wait for _all_ promises to complete
  const results = await Promise.all(promises);
  // sum and return
  return results.reduce((acc, n) => (acc + n), 0);
}

function runObservable(url: string, count: number): Promise<number> {
  // construct {count} observables from the {count} promises
  // (similar to runAsync)
  const requests: Array<Observable<number>> = Array(count).fill(url)
    .map((u) => from(fetchEndpoint(u)));
  // merge Array<Observable<number>> => Observable<Array<number>>
  // ... => Observable<number> containing sum total bytes
  const reduced = merge(...requests).pipe(reduce((acc, x) => acc + x));
  // to Promise for async function
  return firstValueFrom(reduced);
}

const RequestCount: number = 100;

if (process.argv.length < 3) {
  const myname = path.basename(process.argv[1]);
  console.log(`Usage: ${myname} URL`);
  process.exit(1);
}
const TargetUrl: string = process.argv[2];

// map operation to label for data-driven operation
const operationMap: Map<SomeOperation, string> = new Map([
  [runSync, "Sync"],
  [runAsync, "Async"],
  [runObservable, "Observable"],
]);

operationMap.forEach((label, f) => {
  console.log(`${label} starting...`);

  const startTime: number = Date.now();
  f(TargetUrl, RequestCount).then((bytes) => {
    const delta = Date.now() - startTime;
    console.log(`${label}: ${bytes} bytes in ${delta}ms.`);
  });
});
